Remove ImplicitTensorToNum

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32761 Remove ImplicitTensorToNum**
* #32760 remove unnecessary Node* ops
* #32682 Remove Node dependencies from operator.h

This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

zdevito committed:
Remove ImplicitTensorToNum

This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 581955a3:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=581955a3c30983326aa82ebf5977fa5005b9871d).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4367927) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92469944) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92469944))
<details>
<summary>
<code>Jan 29 04:54:03 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 29 04:54:03 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 29 04:54:03 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 29 04:54:03 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 29 04:54:03 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 29 04:54:03 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 29 04:54:03 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 29 04:54:03 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 29 04:54:03 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 29 04:54:03 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 29 04:54:03 processing existing schema:  prim::Constant() -> (...) 
Jan 29 04:54:03 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 29 04:54:03  
Jan 29 04:54:03 Broken ops: [ 
Jan 29 04:54:03 	prim::Drop(...) -> (...) 
Jan 29 04:54:03 	prim::Store(...) -> (...) 
Jan 29 04:54:03 	prim::Load(...) -> (...) 
Jan 29 04:54:03 	prim::ImplicitTensorToNum(Tensor a) -> (Scalar) 
Jan 29 04:54:03 	aten::_ncf_view(Tensor self, int[] input_shape, int normalized_ndim) -> (Tensor) 
Jan 29 04:54:03 	aten::_ncf_unsqueeze(Tensor self, int ndim) -> (Tensor) 
Jan 29 04:54:03 ] 
Jan 29 04:54:03 + cleanup 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito committed:
Update on "Remove ImplicitTensorToNum"


This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

[ghstack-poisoned]

zdevito committed:
Update on "Remove ImplicitTensorToNum"


This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

[ghstack-poisoned]

bhosmer commented on torch/csrc/jit/script/schema_matching.cpp:
```diff
@@ -88,22 +88,22 @@ Value* tryConvertToType(
+    } else if (value_equals_number && concrete_float) {
+      value = graph.insert(aten::Float, {value});
+    } else if (value_equals_number && concrete_int) {
+      value = graph.insert(aten::Int, {value});
```
nit: I think this might be more readable with nesting, no?

bhosmer reviewed (approved):
yesssss

zdevito committed:
Update on "Remove ImplicitTensorToNum"


This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

[ghstack-poisoned]

zdevito committed:
Update on "Remove ImplicitTensorToNum"


This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

[ghstack-poisoned]

zdevito committed:
Update on "Remove ImplicitTensorToNum"


This replaces ImplicitTensorToNum with result-specific operators like
IntImplicit, FloatImplicit, or ScalarImplicit. Note that ScalarImplicit
was not correctly implemented before and this PR fixes the lapse.

This does not change on-disk serialization because these operators are not
serialized directly but written as eg. `annotated(int, foo)`.

Differential Revision: [D19615385](https://our.internmc.facebook.com/intern/diff/D19615385)

[ghstack-poisoned]
